Add Timeout for scanner pod

Vincent056 opened the pull request:
The scan now has a timeout option that can be specified in the `ComplianceScanSettings `object as a duration string (e.g. 1h30m). If the scan does not finish within the specified timeout, it will either be reattempted (up to a maximum of `MaxRetryOnTimeout` times) or considered a failure, depending on the value of `MaxRetryOnTimeout`. The timeout can be disabled by setting it to 0s, and the default value is 30m. The default value for `MaxRetryOnTimeout` is 3, so the timeout scan will be retried up to three times if it fails.

To set a `Timeout` and `MaxRetryOnTimeout` in ScanSetting:

```
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSetting
metadata:
  name: default
  namespace: openshift-compliance
rawResultStorage:
  rotation: 3
  size: 1Gi
roles:
- worker
- master
scanTolerations:
- effect: NoSchedule
  key: node-role.kubernetes.io/master
  operator: Exists
schedule: '0 1 * * *'
timeout: '10m0s'
maxRetryOnTimeout: 3
```

So the idea is to check the creation timestamp of the Scanner pod age, if it is longer than timerout, we will terminate the scan or retry.

A timeout scan will send an event warning on retries, and the scan will have an error result.

jhrozek commented on pkg/apis/compliance/v1alpha1/compliancescan_types.go:
```diff
@@ -204,6 +205,10 @@ type ComplianceScanSettings struct {
+
+	// Timeout is the maximum amount of time the scan can run. If the scan
+	// hasn't finished by then, it will be aborted.
+	Timeout string `json:"timeout,omitempty"`
```
As discussed yesterday on a call, this should use a kubebuiler default (https://book.kubebuilder.io/reference/markers/crd-validation.html)

jhrozek commented on pkg/controller/compliancescan/aggregator.go:
```diff
@@ -139,7 +140,6 @@ func (r *ReconcileComplianceScan) deleteAggregator(instance *compv1alpha1.Compli
-
 	podName := getAggregatorPodName(scanInstance.Name)
-	return isPodRunning(r, podName, common.GetComplianceOperatorNamespace(), logger)
+	return isPodRunning(r, podName, common.GetComplianceOperatorNamespace(), time.Duration(0), logger)
```
is the meaning of time.Duration(0) "no timeout"? In that case we might want to add a global variable with that name for better readability

Vincent056 commented:
> I added a bunch of comments but in the middle of a review I realized that Vincent mentioned some pod timeout option, another option might be to use a job to kill the pods after some time. Anyway, Vincent, let me know if this PR is ready for review or still a WIP

Just updated the code, and it should be ready for a review, not sure if is it a good way to do the e2e test, besides that, I added a retry option, and I used the creation timestamp to determine the timeout instead of log output.

jhrozek commented on config/manager/kustomization.yaml:
```diff
@@ -3,7 +3,7 @@ resources:
 images:
 - name: compliance-operator
-  newName: quay.io/compliance-operator/compliance-operator
+  newName: image-registry.openshift-image-registry.svc:5000/openshift/compliance-operator
```
this was probably commited by mistake

JAORMX replied:
oh, yep, gotta be changed.

Vincent056 replied:
oops

jhrozek commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -209,6 +213,9 @@ func (r *ReconcileComplianceScan) validate(instance *compv1alpha1.ComplianceScan
 	// If no phase set, default to pending (the initial phase):
 	if instance.Status.Phase == "" {
 		instanceCopy := instance.DeepCopy()
+		if instance.Status.RemainingRetries != instance.Spec.MaxRetryOnTimeout {
```
Do we need this if? Wouldn't this work equally well if we just set `instanceCopy.Status.RemainingRetries = instance.Spec.MaxRetryOnTimeout` ?

Vincent056 replied:
Yes, it would be the same

jhrozek commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -288,6 +295,13 @@ func (r *ReconcileComplianceScan) phasePendingHandler(instance *compv1alpha1.Com
 		return reconcile.Result{}, err
 	}
 
+	if _, hasTimeoutAnn := instance.Annotations[compv1alpha1.ComplianceScanTimeoutAnnotation]; hasTimeoutAnn {
```
This is a nit, but I think it would read nicer if we could wrap this in a method like `NeedsRescan`, maybe we could even share the body of the methods and just pass the name of the annotation into the inner method.

Vincent056 replied:
that's a good suggestion, I will use a method here

jhrozek reviewed (commented):
I'll continue the review later.

xiaojiey commented:
/hold for QE review

Vincent056 commented:
/retest

jhrozek commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -99,10 +102,11 @@ var _ reconcile.Reconciler = &ReconcileComplianceScan{}
-	Recorder record.EventRecorder
-	Metrics  *metrics.Metrics
+	Client    client.Client
+	ClientSet *kubernetes.Clientset
```
`ClientSet` seems to be unused now?

jhrozek commented on pkg/controller/compliancescan/scantype.go:
```diff
@@ -145,21 +146,33 @@ func (nh *nodeScanTypeHandler) createScanWorkload() error {
+	timeoutNodes := []string{}
+	var err error
+	if nh.scan.Spec.ComplianceScanSettings.Timeout != "" {
+		timeoutVal, err = time.ParseDuration(nh.scan.Spec.ComplianceScanSettings.Timeout)
```
I wonder if this should be moved to `ReconcileComplianceScan.validate()`

Vincent056 replied:
The timeout value is being passed to isPodRunningInNode, wondering what should I do here?

jhrozek replied:
(sorry for the late reply)
I would say parse twice, once in `validate` and once before use. The reason is that if the duration fails parsing on validate we would go straight to `DONE/ERROR` whereas if the parsing fails just before use, the reconcile loop would retry indefinitely.

jhrozek replied:
@Vincent056 all my comments about the PR have been addressed except for this one - let me know if you disagree and I should just LGTM the patch. Otherwise great work!

Vincent056 replied:
Sorry for the delay, it has been fixed now!

jhrozek reviewed (changes requested):
Three more nits (plus I think one refactoring ask was still there from last time)

Vincent056 commented on CHANGELOG.md:
```diff
@@ -9,7 +9,12 @@ Versioning](https://semver.org/spec/v2.0.0.html).
+  specified in the `ComplianceScanSettings` object as a duration string,
+  e.g. `1h30m`. If the scan is not completed within the timeout, the
+  corresponding scan will either fail or retry.
+  See [PR](https://github.com/ComplianceAsCode/compliance-operator/pull/192) for details usage.
```
that's a good point, just added there as well

jhrozek replied:
Thank you, there is one typo that I'll point out separately, but this is much better.

xiaojiey commented:
PR verification with 4.13.0-0.nightly-2023-01-11-225449+code in the PR
Test Steps
1. deploy complaince operator with the PR
2. Create ss and ssb:
$ cat non_default_timeout_2.yaml 
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSetting
maxRetryOnTimeout: 3
metadata:
  name: non-default-timeout-2
  namespace: openshift-compliance
rawResultStorage:
  nodeSelector:
    node-role.kubernetes.io/master: ""
  pvAccessModes:
  - ReadWriteOnce
  rotation: 3
  size: 1Gi
  tolerations:
  - effect: NoSchedule
    key: node-role.kubernetes.io/master
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  - effect: NoSchedule
    key: node.kubernetes.io/memory-pressure
    operator: Exists
roles:
- master
- worker
scanTolerations:
- operator: Exists
schedule: 0 1 * * *
showNotApplicable: false
strictNodeScan: true
timeout: "10s"
$ oc apply -f non_default_timeout_2.yaml 
scansetting.compliance.openshift.io/non-default-timeout-2 created
$ oc compliance bind -N timeout2 -S non-default-timeout-2 profile/ocp4-cis profile/ocp4-cis-node
Creating ScanSettingBinding timeout2
$ oc get scan -w
NAME                   PHASE       RESULT
ocp4-cis               RUNNING     NOT-AVAILABLE
ocp4-cis-node-master   LAUNCHING   NOT-AVAILABLE
ocp4-cis-node-worker   RUNNING     NOT-AVAILABLE
ocp4-cis-node-master   RUNNING     NOT-AVAILABLE
ocp4-cis               RUNNING     NOT-AVAILABLE
ocp4-cis-node-master   RUNNING     NOT-AVAILABLE
ocp4-cis               DONE        ERROR
ocp4-cis-node-master   DONE        ERROR
ocp4-cis-node-worker   RUNNING     NOT-AVAILABLE
ocp4-cis-node-worker   DONE        ERROR
^C
$ oc get scan  ocp4-cis  -o=jsonpath={.metadata.annotations}
{"compliance.openshift.io/timeout":"api-checks"} 
$ oc get scan ocp4-cis-node-master -o=jsonpath={.metadata.annotations}
{"compliance.openshift.io/timeout":"ip-10-0-142-203.us-east-2.compute.internal"}
$ oc get scan ocp4-cis-node-worker -o=jsonpath={.metadata.annotations}
{"compliance.openshift.io/timeout":"ip-10-0-214-228.us-east-2.compute.internal"}

xiaojiey commented:
/label qe-approved

xiaojiey commented:
/unhold

jhrozek commented on doc/usage.md:
````diff
@@ -531,3 +531,42 @@ $ oc patch sub compliance-operator -nopenshift-compliance --patch-file co-memlim
+```
+
+Compliance Operator will check the creation timestamp of the Scanner pod age,
+if it is longer than timerout, we will terminate the scan or retry.
````
typo: timerout

jhrozek commented on doc/usage.md:
```diff
@@ -531,3 +531,42 @@ $ oc patch sub compliance-operator -nopenshift-compliance --patch-file co-memlim
+Compliance Operator will check the creation timestamp of the Scanner pod age,
+if it is longer than timerout, we will terminate the scan or retry.
+
+A timeout scan will send an event warning on retries, and the scan will have an
```
A warning even would read better then an event warning, but I'd defer checking English to @rhmdnd

jhrozek commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -375,11 +387,22 @@ func (r *ReconcileComplianceScan) phaseLaunchingHandler(h scanTypeHandler, logge
+	if len(timeoutNodes) > 0 {
+		scan := h.getScan()
+		scan = scan.DeepCopy()
+		if _, hasTimeoutAnnotation := scan.Annotations[compv1alpha1.ComplianceScanTimeoutAnnotation]; hasTimeoutAnnotation {
```
Could we use `scan.NeedsTimeoutRescan` here?

jhrozek commented on config/rbac/operator_cluster_role.yaml:
```diff
@@ -101,6 +101,7 @@ rules:
       - configmaps
       - pods
       - secrets
+      - pods/log
```
do we still need this permission?

jhrozek reviewed (changes requested):
Some more nits added. I'm sorry this review is taking so long, but the PR is quite complex. Thanks for the patience!

jhrozek commented:
The only issue that's unclear if we want to have it resolved or not is the [one about validation](https://github.com/ComplianceAsCode/compliance-operator/pull/192#discussion_r1071302482) otherwise I'll be happy to add LGTM

jhrozek reviewed (approved):
/lgtm
thank you for the patience

Vincent056 commented:
> /lgtm thank you for the patience

thanks for all the reviews!

mkumku commented:
This looks straight forward to me.
I do not believe TE is needed for this, since it will be in release notes/documentaion, right @sheriff-rh ?

Vincent056 commented:
/test all
